drivers/sensor/bmi160/bmi160_v2.c

```diff
@@ -0,0 +1,151 @@
+/* Copyright (c) 2022 Google Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <string.h>
+
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/kernel.h>
+#include <zephyr/math/util.h>
+#include <zephyr/sys_clock.h>
+
+#include "bmi160.h"
+
+static int bmi160_read_data(const struct device *sensor, struct sensor_data *out)
+{
+	struct bmi160_data *data = sensor->data;
+	int rc;
+	uint16_t *sample;
+
+	out->header.base_timestamp = k_uptime_get() * USEC_PER_MSEC;
+
+	if ((rc = bmi160_sample_fetch(sensor)) != 0) {
+		return rc;
+	}
+
+	for (int i = 0; i < out->header.reading_count; ++i) {
+		switch (out->readings[i].sensor_type) {
+		case SENSOR_TYPE_ACCELEROMETER:
```

Comment:
Why have this at all, why not let a user specific which channels they want mapped to which numerical values and use a function pointer to interpret a sample into the value

Reply:
I think this is similar to what Gwendal was saying. This takes us back to my original implementation which had different `struct sensor_<type>_data`. The various APIs passed in a `void *` and a `sensor_type`, then the sensor would be able to cast the `void *` to the appropriate pointer based on the type. I will likely not be able to get this finished tonight so I'll start the discussion on this thread.

The `sensor_read_data` API changed with the request to fetch multiple channels at the same time without using the FIFO. To do that I used the `struct sensor_data` as both an input and an output. The `header.reading_count` told me how many samples we're asking the driver to fetch, while `readings[?].sensor_type` told the driver which sample to fetch. The driver is then free to optimize and fetch all the samples, buffer them, then fill the appropriate `readings[]` elements while throwing away the remaining samples if any. I don't like this approach and I'm wondering if this requirement is really a hard requirement. It seems to me that the `sensor_read_data` API which is blocking should never be in a tight pipeline. Meaning that if the application needed fast samples, they should be using the FIFO. If this is true, then we can go back to the signature that accepts the `sensor_type` and this simplifies the API because we can now pass a `void *` as the `data` argument and cast it depending on the type.